vendor/github.com/rancher/norman/controller/generic_controller.go

```diff
@@ -124,26 +124,37 @@ func (g *genericController) Enqueue(namespace, name string) {
 
 func (g *genericController) AddHandler(ctx context.Context, name string, handler HandlerFunc) {
 	g.Lock()
+	defer g.Unlock()
+
+	g.generation++
 	h := &handlerDef{
 		name:       name,
 		generation: g.generation,
 		handler:    handler,
 	}
-	g.handlers = append(g.handlers, h)
-	g.Unlock()
 
-	go func() {
+	go func(gen int) {
 		<-ctx.Done()
 		g.Lock()
-		var handlers []*handlerDef
+		defer g.Unlock()
+		var newHandlers []*handlerDef
 		for _, handler := range g.handlers {
-			if handler != h {
-				handlers = append(handlers, h)
+			if handler.generation == gen {
+				continue
 			}
+			newHandlers = append(newHandlers, handler)
+		}
+		g.handlers = newHandlers
+	}(h.generation)
+
+	g.handlers = append(g.handlers, h)
```

Comment:
Trying to wrap my head around this. It looks like we run through this sequentially and with the locking there is no way we could run through this sequentially. 

So, first time we run through the function we 
1. Acquire the lock 
2. Add the handlers to `g.Handlers` (see above) 
3. Check for sync
4. Unlock
4. Now, we `goto` line 138 and acquire the lock 
5. We iterate over the list of handlers skipping those that have the same generation
6. We overwrite `g.handlers` on  line 147 with handlers whose `generation != gen`

I think this could be simplified out. I'm not sure why we are using the goroutine here if it cannot execute concurrently due to the locking.

Reply:
https://play.golang.org/p/w6vzffRUqkl for testing